includes/abilities-api/class-wp-ability-categories-registry.php

```diff
@@ -0,0 +1,254 @@
+<?php
+/**
+ * Abilities API
+ *
+ * Defines WP_Ability_Categories_Registry class.
+ *
+ * @package WordPress
+ * @subpackage Abilities API
+ * @since 6.9.0
+ */
+
+declare( strict_types = 1 );
+
+/**
+ * Manages the registration and lookup of ability categories.
+ *
+ * @since 6.9.0
+ * @access private
+ */
+final class WP_Ability_Categories_Registry {
+	/**
+	 * The singleton instance of the registry.
+	 *
+	 * @since 6.9.0
+	 * @var self|null
+	 */
+	private static $instance = null;
+
+	/**
+	 * Holds the registered ability categories.
+	 *
+	 * @since 6.9.0
+	 * @var WP_Ability_Category[]
+	 */
+	private $registered_categories = array();
+
+	/**
+	 * Registers a new ability category.
+	 *
+	 * Do not use this method directly. Instead, use the `wp_register_ability_category()` function.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @see wp_register_ability_category()
+	 *
+	 * @param string               $slug The unique slug for the ability category. Must contain only lowercase
+	 *                                   alphanumeric characters and dashes.
+	 * @param array<string, mixed> $args {
+	 *     An associative array of arguments for the ability category.
+	 *
+	 *     @type string               $label       The human-readable label for the ability category.
+	 *     @type string               $description A description of the ability category.
+	 *     @type array<string, mixed> $meta        Optional. Additional metadata for the ability category.
+	 * }
+	 * @return WP_Ability_Category|null The registered ability category instance on success, null on failure.
+	 */
+	public function register( string $slug, array $args ): ?WP_Ability_Category {
+		if ( $this->is_registered( $slug ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: Ability category slug. */
+				sprintf( __( 'Ability category "%s" is already registered.' ), esc_html( $slug ) ),
+				'6.9.0'
+			);
+			return null;
+		}
+
+		if ( ! preg_match( '/^[a-z0-9]+(?:-[a-z0-9]+)*$/', $slug ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'Ability category slug must contain only lowercase alphanumeric characters and dashes.' ),
+				'6.9.0'
+			);
+			return null;
+		}
```

Comment:
Are we not checking for the category init hook anymore?

Reply:
That's interesting! I didn't realize we were using subsequent hooks as a way of narrowing the registered abilities by context. I don't disagree, it just wasn't a means of "filtering" I'd considered, yet. My assumption so far was that, yes, _all_ abilities would be registered and then subsequent filtering would occur. There are certainly trade-offs of both approaches.